basic/model.py

```diff
@@ -68,143 +96,180 @@ def __init__(self, config, scope, rep=True):
-
-            (fw_g0, bw_g0), _ = bidirectional_dynamic_rnn(first_cell, first_cell, p0, x_len, dtype='float', scope='g0')  # [N, M, JX, 2d]
-            g0 = tf.concat(3, [fw_g0, bw_g0])
-            (fw_g1, bw_g1), _ = bidirectional_dynamic_rnn(first_cell, first_cell, g0, x_len, dtype='float', scope='g1')  # [N, M, JX, 2d]
-            g1 = tf.concat(3, [fw_g1, bw_g1])
-
-            logits = get_logits([g1, p0], d, True, wd=config.wd, input_keep_prob=config.input_keep_prob,
-                                mask=self.x_mask, is_train=self.is_train, func=config.answer_func, scope='logits1')
-            a1i = softsel(tf.reshape(g1, [N, M * JX, 2 * d]), tf.reshape(logits, [N, M * JX]))
-            a1i = tf.tile(tf.expand_dims(tf.expand_dims(a1i, 1), 1), [1, M, JX, 1])
-
-            (fw_g2, bw_g2), _ = bidirectional_dynamic_rnn(d_cell, d_cell, tf.concat(3, [p0, g1, a1i, g1 * a1i]),
-                                                          x_len, dtype='float', scope='g2')  # [N, M, JX, 2d]
-            g2 = tf.concat(3, [fw_g2, bw_g2])
-            logits2 = get_logits([g2, p0], d, True, wd=config.wd, input_keep_prob=config.input_keep_prob,
-                                 mask=self.x_mask,
+                p0 = attention_layer(config, self.is_train, encoded_passage, encoded_question, h_mask=self.passage_mask, u_mask=self.question_mask, scope="p0", tensor_dict=self.tensor_dict)
+                first_cell = cell_with_dropout
+
+            # The modeling layer
+            # The 0th (first) bidirectional encoder in the modeling layer
+            # [batch_size, max_num_sentences, max_sentence_size, 2*hidden_size]
+            (fw_output_biencoder0, bw_output_biencoder0), _ = bidirectional_dynamic_rnn(first_cell, first_cell, p0, passage_len, dtype='float', scope='model_layer_biencoder_0')
+            modeled_passage_0 = tf.concat(3, [fw_output_biencoder0, bw_output_biencoder0])
+
+            # The 1st (second) bidirectional encoder in the modeling layer
+            # [batch_size, max_num_sentences, max_sentence_size, 2*hidden_size]
+            (fw_output_biencoder1, bw_output_biencoder1), _ = bidirectional_dynamic_rnn(first_cell, first_cell, modeled_passage_0, passage_len, dtype='float', scope='model_layer_biencoder_1')
+            # This encoding is used to get the start span index.
+            modeled_passage_1 = tf.concat(3, [fw_output_biencoder1, bw_output_biencoder1])
+
+            # The logits for the start span answer.
+            logits = get_logits([modeled_passage_1, p0], hidden_size, True, wd=config.wd, input_keep_prob=config.input_keep_prob,
+                                mask=self.passage_mask, is_train=self.is_train, func=config.answer_func, scope='logits1')
+
+            # TODO: NOT QUITE SURE WHAT'S GOING ON HERE EITHER
+            a1i = softsel(tf.reshape(modeled_passage_1, [batch_size, max_num_sentences * max_sentence_size, 2 * hidden_size]), tf.reshape(logits, [batch_size, max_num_sentences * max_sentence_size]))
```

Comment:
Min's `a1i` is my `passage_weighted_by_predicted_span`, here: https://github.com/allenai/deep_qa/blob/94177eb084fb70fa198e216a05f77e335d388ea0/deep_qa/models/reading_comprehension/bidirectional_attention.py#L182-L185.